Code challenge

addiebarron opened the pull request:
## Changes
* Implements the `AddressParse` view to handle API calls and parse addresses
&rarr; `parserator_web/views.py`
* Adds view unit tests
&rarr; `tests/test_views.py`
* Adds a JS frontend component to send API calls and display response data
&rarr; `parserator_web/static/js/index.js`

## Testing
To run the Python and Javascript test suites (which include linting), run the following from the project root:
```
docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app
```

## Building and running
First, make sure you have [Docker](https://docs.docker.com/install/) and [Docker Compose](https://docs.docker.com/compose/install/) installed on your system. 

To build the containers, run `docker-compose build`. 

To run the app, run `docker-compose up`.  App logs will begin output to your terminal, and the web interface will be accessible at http://localhost:8000. 

You can optionally combine these two steps with `docker-compose up --build`.

addiebarron committed:
Backend views and frontend complete

addiebarron committed:
Tests first pass

addiebarron committed:
Linting fixes

addiebarron committed:
All tests passing, js cleanup

derekeder commented on tests/conftest.py:
```diff
@@ -1 +1,7 @@
-# Define test fixtures here.
+# import pytest
```
Since these lines are commented out, we should probably remove them

addiebarron replied:
Makes sense, thanks. Fixed in https://github.com/addiebarron/datamade-code-challenge-2021/pull/1/commits/cfbd9e0575fee84a380d15ed92ebf977e2eed465.

derekeder commented on .eslintrc.js:
```diff
@@ -10,6 +10,9 @@ module.exports = {
+    "space-before-blocks": ["error", "always"],
+  },
+  parserOptions: {
+    ecmaVersion: "2017",
```
I appreciate adding this specification for Javascript linting!

addiebarron replied:
Thanks! I wasn't able to pass the tests without it ... and I was too stubborn to rewrite everything in ES5. I would probably want to make sure my style & syntax are consistent with the team's before making a change like this in practice.

derekeder commented on tests/test_views.py:
```diff
@@ -1,15 +1,48 @@
+
+    assert response.status_code == 200 and data['status'] == 200
+    # Is there a better way to test timestamp format? Is it necessary?
+    assert type(error['timestamp']) == str
```
I'd say it's probably not necessary as the timestamp is not really important to the output of parserator (though it is helpful info for the user). I'd probably remove this assertion

derekeder commented on parserator_web/views.py:
```diff
@@ -14,11 +14,65 @@ class AddressParse(APIView):
+            response_data = {
+                'status': status,
+                'error': {
+                    'timestamp': timestamp,
```
nice addition of the timestamp! I can see it being helpful for users of the API service to track and log their usage.

addiebarron replied:
I took the tip from [this article](https://www.baeldung.com/rest-api-error-handling-best-practices) that Hannah shared with me during our last code review!

derekeder reviewed (commented):
Hi @addiebarron - This is really strong code - thanks for submitting it. Overall its very clean and you cover all the potential input strings nicely.

I just have a few inline comments on cleanup and documentation we'd like you to address. Additionally, could you update the description of this PR to include instructions for building and running the app, as well as any high-level notes you'd like to pull out from your inline code comments?

addiebarron committed:
Address code review from Derek

addiebarron commented:
Hi @derekeder, thanks for this helpful review, and for your kind words. I've removed the lines you flagged in `tests/test_views.py` and added a "Building and running" section to the PR description. The code changes are in https://github.com/addiebarron/datamade-code-challenge-2021/pull/1/commits/cfbd9e0575fee84a380d15ed92ebf977e2eed465. Let me know if there are any other changes you'd like me to make before our chat tomorrow. Looking forward!

derekeder commented:
Great - thank you!
